flink-runtime/src/main/java/org/apache/flink/runtime/security/modules/HadoopModule.java

```diff
@@ -97,8 +97,7 @@ public void install() throws SecurityInstallException {
                     // If UGI use keytab for login, do not load HDFS delegation token.
                     for (Token<? extends TokenIdentifier> token : usrTok) {
                         if (!token.getKind().equals(hdfsDelegationTokenKind)) {
-                            final Text id = new Text(token.getIdentifier());
-                            credentialsToBeAdded.addToken(id, token);
+                            credentialsToBeAdded.addToken(token.getService(), token);
```

Comment:
@lirui-apache Let me explain this a little bit.

1) HDFS-9276 made a defensive copy because a `Token` instance is passed to the constructor of `PrivateToken`. In the view of the `PrivateToken` class, a defensive copy is certainly needed.
2) For this change, `token`, along with its referencing object `usrTok` and `credentialsFromTokenStorageFile`, are all local variables. `credentialsFromTokenStorageFile` comes from reading a file. Therefore, a deep copy is not as necessary here as it is in HDFS-9276.
3) `token.getIdentifier()` returns a `byte[]`, while `token.getService()` returns a `Text`. This is probably why a deep copy looks like removed in this change. The fact is that previously to this change, it didn't make a deep copy either.

Reply:
Hey @ChengbingLiu , thanks for the explanations. I'm not very familiar with this topic. So I have some follow-up questions.
1. Per HDFS-9276, the deep copy is added to avoid scenarios like `"Whenever the content of token.service is changed, publicService will reflect the change. I don't think this is what we want."` So I wonder whether we need to guard against similar situations. E.g. when `token.service` is changed, do we want it to be reflected in the `Credentials::tokenMap`?
2. The `usrTok` is indeed local in `HadoopModule`. However, a similar change is made in `Utils::setTokensFor` where `usrTok` is retrieved from current UGI and therefore is not local to the method. Will that be an issue for us?
3. It seems both `Text(byte[])` and `Text(Text)` copy the byte array. So previously to this change, we do make a deep copy, no?

Reply:
> E.g. when `token.service` is changed, do we want it to be reflected in the `Credentials::tokenMap`?

No, we don't want that. But that should not happen with this change, explained below.

> However, a similar change is made in `Utils::setTokensFor` where `usrTok` is retrieved from current UGI and therefore is not local to the method. Will that be an issue for us?

I don't think so. In `Utils::setTokensFor`, all tokens retrieved from the UGI is used to construct the local variable `credentials`, which is then used to be serialized into `ByteBuffer securityTokens`, which will not reflect the change in the UGI.

> It seems both `Text(byte[])` and `Text(Text)` copy the byte array. So previously to this change, we do make a deep copy, no?

Sorry for not making myself clear. I think the purpose of the previous code `new Text(token.getIdentifier())` was type conversion instead of deep copy, otherwise it should look like `addToken(new Text(...), new Token(token))`.

Reply:
I see. Since `credentials` is serialized right away, I'm fine with this change.

Reply:
Thanks my mentor @ChengbinLiu for the explanations. Thanks for @lirui-apache review.